### PR TITLE
libexpr: add function serialization builtins

### PR DESCRIPTION
**Pull request by KiaraGrouwstra**
## Summary

Add `builtins.serializeFunction` and `builtins.deserializeFunction` behind a new `function-serialization` experimental feature flag, enabling round-trip serialization of Nix functions including their closures.

Also extends `builtins.toString` to coerce functions to strings.

Addresses NixOS/nix#15049.

Disclaimer I used a coding agent in the creation of this patch.

/cc @roberth @Ericson2314

## Motivation

Function serialization enables:
- Debugging and logging of function values
- Dynamic code generation of `.nix` files
- Bundling eval-time logic + captured configuration into a single string for evaluation continuations (e.g. dynamic derivations via RFC 0092)

## Design

`builtins.serializeFunction` produces a self-contained Nix expression string. Closure variables are reconstructed as recursive `let` bindings. The result can be passed through `builtins.deserializeFunction` to recover a callable function.

```nix
let
  x = 1;
  serialized = builtins.serializeFunction (y: x + y);
  restored = builtins.deserializeFunction serialized;
in restored 5  # => 6
```

An attrset form provides additional control:

```nix
builtins.serializeFunction {
  fn = myFunction;
  allowedPluginPrimOps = [ "getFlake" ];
  preserveStringContext = true;
  externalBindings = [ "concatMapStringsSep" ];
}
```

## What works

- Lambdas with full closure reconstruction (nested closures, with-bound variables, recursive values via back-referenced let bindings)
- Primops as `builtins.<name>`, partially applied primops with arguments preserved
- Eta-reduction normalizes lib polyfill wrappers (`x: builtins.head x` -> `builtins.head`; also handles direct builtin selects and multi-arg wrappers)
- Plugin primop control via `allowedPluginPrimOps` (rejected by default, configurable allowlist)
- String context preservation via `preserveStringContext` (emits `builtins.appendContext` calls)
- External bindings via `externalBindings` (library functions emitted as wrapper parameters instead of inlined, keeping output small and nixpkgs-version-independent)
- derivation and other base environment names work after deserialization

## Known limitations

- String contexts lost by default (opt-in via `preserveStringContext`)
- Serialized form not stable across Nix versions
- Float precision limited by `std::to_string`
- `deserializeFunction` evaluates arbitrary expressions (inherent to approach)

Full details in the `function-serialization` experimental feature description.

## Test plan

- [x] 52 new unit tests covering all serialization features, edge cases, and dynamic derivation scenario tests
- [x] `nix eval --extra-experimental-features function-serialization --expr 'let x = 1; in builtins.serializeFunction (y: x + y)'`
- [x] `nix eval --extra-experimental-features function-serialization --expr 'let f = builtins.deserializeFunction "(x: x + 1)"; in f 41'`
- [x] Verify no test regressions without the feature enabled

### CODE REVIEW

**Comment by xokdvium** [comment]
OTOH, I think since Nix doesn't have referential transparency (functions and values in general have pointer identity which is impossible to preserve without also serialising the heap), it's impossible to faithfully implement.

File: src/libexpr/primops.cc [hunk 1]
```diff
@@ -4493,10 +4494,847 @@ static RegisterPrimOp primop_toString({
...
+          Lambdas produce their source representation (e.g. `toString (x: x)` yields `"(x: x)"`).
+          Primops produce their qualified `builtins.` name
+          (e.g. `toString builtins.head` yields `"builtins.head"`).
+          Partially applied primops include their arguments
+          (e.g. `toString (builtins.map builtins.head)` yields `"(builtins.map builtins.head)"`).
+    )doc",
     .impl = prim_toString,
 });
 
+/**
```
Comment (edolstra) [inline src/libexpr/primops.cc:4508]: (not commenting on the general idea of this PR)

`primops.cc` is already way too big, please add new primops to `src/libexpr/primops/`.

**Comment by edolstra** [comment]
The complexity of this function makes it quite tricky to have as a builtin, given the goal of language reproducibility. (I.e. any two versions of Nix should serialize a function in the same way, since otherwise you can end up with different store paths.) E.g. `from,to{YAML}` was considered too risky for that reason.

Maybe this could be done in WASM using `builtins.wasm`. However, currently `builtins.wasm` doesn't expose the `Expr` AST. And we really don't want to make the AST part of the stable interface, since it does change over time.
